wapitiCore/attack/mod_methods.py

```diff
@@ -56,24 +108,49 @@ async def attack(self, request: Request, response: Optional[Response] = None):
         log_verbose(f"[+] {option_request}")
 
         try:
-            response = await self.crawler.async_send(option_request)
+            option_response = await self.crawler.async_send(option_request)
         except RequestError:
             self.network_errors += 1
             return
 
-        if response.is_success or response.is_redirect:
-            methods = response.headers.get("allow", '').upper().split(',')
-            methods = {method.strip() for method in methods if method.strip()}
-            interesting_methods = sorted(methods - self.KNOWN_METHODS)
-
-            if interesting_methods:
-                log_orange("---")
-                log_orange(f"Interesting methods allowed on {page}: {', '.join(interesting_methods)}")
-                await self.add_addition(
-                    category=NAME,
-                    request=option_request,
-                    info=f"Interesting methods allowed on {page}: {', '.join(interesting_methods)}",
-                    wstg=WSTG_CODE,
-                    response=response
-                )
-                log_orange("---")
+        if not (option_response.is_success or option_response.is_redirect):
+            return
```

Comment:
If we do this, we should ad a log line to say that OPTIONS has no response, and that we are going to try an hardcoded list of methods.

Reply:
@Slokilla I think you mean 405 ?

I think the logic should be changed in the following way:

- if `OPTIONS` returns the `Allow` header, rely solely on it
- if it returns a 405 test each method separately, ignore 403 and 405 responses otherwise compare responses

wdyt?

Reply:
Here is a case I saw on a CTF where `OPTIONS` returned a 405 but `PUT` was allowed : https://devl00p.github.io/posts/Solution-du-CTF-Rubbish-de-Wizard-Labs/#claude-rish

Reply:
Actually, I mean anything that differ from 200 because the way that OPTIONS failed isn't that important.

So i would modify a bit your logic to: 
  - if OPTIONS returns the _Allow_ header, rely solely on it
  - if it fails test each method separately [...]

I hope I'm not talking too much nonsense ^^

Reply:
If we are heading that way, shouldn't we start considering both 404 and 405 status code, since 404 will indeed fall into the exception ? Moreover, a server can totally customize the way he can answer to any type of request, so I'm confused regarding the type of behavior that can be relevant to us. 

I mean, any unexpected returned headers might be done on purpose or results from an unexpected behavior from the server. 
We should see every other return code similar to 405 and 404 returned by the ``OPTIONS`` method knowing that the method module will try to see differences between all the methods (or at least, the one specified in the ``Allow`` header with the ``OPTIONS`` method) and the get request made by the crawler.

Plus, a ``HEAD`` request is supposed to return the same thing as the ``GET`` request but without the body  (this can also be bypassed with custom headers), however, for now it is still triggered, and I was wondering if creating a special case for that behavior could be a good idea.

What are your thoughts on those points ?

Reply:
For the OPTIONS response indeed I think we can rely only on the Allow header whatever the status code.

Then for custom methods, ignore 400/401/403/404/405 responses ? I think the status code is explicit.

Maybe we can also check for false positives using the status code returned by an invalid method ?

```console
$ curl -s -D- -XOPTIONS https://talk.jekyllrb.com/ | head -1
HTTP/2 404 
$ curl -s -D- -XPUT https://talk.jekyllrb.com/ | head -1
HTTP/2 404 
$ curl -s -D- -XLOL https://talk.jekyllrb.com/ | head -1
HTTP/2 405 
$ curl -s -D- -XGET https://talk.jekyllrb.com/ | head -1
HTTP/2 200
```
vs

```console
$ curl -s -D- -XGET https://www.tevora.com/ | head -1
HTTP/2 200 
$ curl -s -D- -XOPTIONS https://www.tevora.com/ | head -1
HTTP/2 200 
$ curl -s -D- -XPUT https://www.tevora.com/ | head -1
HTTP/2 200 
$ curl -s -D- -XLOL https://www.tevora.com/ | head -1
HTTP/2 200
```